_drafts/printf-on-embedded.md

```diff
@@ -0,0 +1,566 @@
+
+The biggest downside is you're spending one of your UART peripherals for printf,
+which might be needed for your actual application.
+
+Also, configuring a UART might require managing clock configuration, chip
+power/sleep modes, figuring out buffering the data (DMA?), making sure
+everything is interrupt safe, etc.
+
+And finally, you'll usually need some kind of adapter dongle to interface the
+UART with your PC (eg USB-to-TTL or USB-to-asynchronous serial).
+
+But! because the interface is contained entirely on the microcontroller, it can
+be used without a debug probe, which can make it useful when not actively
+debugging your system (eg you can use it to write log statements to an attached
+PC!).
+
+Worth noting, if your microcontroller has support for USB, you might be able to
+skip the need for an external FTDI-like adapter by implementing CDC
+(Communications Device Class) on-chip. That adds a lot more software to your
+system, but can be useful. Out of scope of this FAQ.
+
+### Semihosting
+
+Next up the food chain, semihosting is a protocol that proxies the "hosting"
+(meaning libc hosting, eg `printf`/`scanf`/`open`/`close` etc) over an attached
+debug probe to the PC.
+
+Some good descriptions:
+
+- <https://wiki.segger.com/Semihosting>
+- <https://developer.arm.com/documentation/dui0375/g/What-is-Semihosting-/What-is-semihosting->
+
+And the reference documentation can be found in:
+
+> ARM Developer Suite (ADS) v1.2 Debug Target Guide, Chapter 5. Semihosting
+
+TLDR, semihosting operates by the target executing the breakpoint instruction
+with a special breakpoint id, `bkpt 0xab`, with the I/O opcode stored in `r1`
+and whatever necessary data stored in a pointer loaded into `r2`.
+
+The debug probe executes the I/O operation based on the opcode and data, then
+resumes the microcontroller.
+
+To use it, the simplest approach is to use newlib's rdimon library spec:
+
+```bash
+# add these to your linker command line
+LDFLAGS += --specs=nano.specs --specs=rdimon.specs
+```
+
+Then add this call somewhere in your system init, prior to calling any libc I/O
+functions:
+
+```c
+// this executes some exchanges with the debug monitor to set up the correct
+// file handles for stdout, stderr, and stdin
+extern void initialise_monitor_handles(void);
+initialise_monitor_handles();
+```
+
+`printf` and friends should work after that! Be sure to check the user manual
+for your debug probe on how it exposes the semihosting interface. For example,
+segger jlink requires running some commands to enable semihosting:
+
+```bash
+# from gdb client
+monitor semihosting enable
+monitor semihosting ioclient 2
+```
+
+(From <https://wiki.segger.com/J-Link_GDB_Server>).
+
+For pyocd, you can pass the `--semihosting` argument when starting the gdb
+server.
+
+As an extra bonus, see below for a simplified semihosting solution based on the
+newlib one that doesn't require linking against `rdimon.specs`, and saves some
```

Comment:
Not sure if I missed, but earlier it was mentioned that we would provide more info on rdimon. Might be good to provide an overview or link before this

Reply:
done!